packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -204,37 +262,60 @@ export class AssetStaging extends Construct {
     return bundleDir;
   }
 
-  private calculateHash(props: AssetStagingProps): string {
-    let hashType: AssetHashType;
-
-    if (props.assetHash) {
-      if (props.assetHashType && props.assetHashType !== AssetHashType.CUSTOM) {
-        throw new Error(`Cannot specify \`${props.assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
+  /**
+   * Determines the hash type from user-given prop values.
+   *
+   * @param assetHashType Asset hash type construct prop
+   * @param assetHash Asset hash given in the construct props
+   */
+  private determineHashType(assetHashType?: AssetHashType, assetHash?: string) {
+    if (assetHash) {
+      if (assetHashType && assetHashType !== AssetHashType.CUSTOM) {
+        throw new Error(`Cannot specify \`${assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
       }
-      hashType = AssetHashType.CUSTOM;
-    } else if (props.assetHashType) {
-      hashType = props.assetHashType;
+      return AssetHashType.CUSTOM;
+    } else if (assetHashType) {
+      return assetHashType;
     } else {
-      hashType = AssetHashType.SOURCE;
+      return AssetHashType.SOURCE;
     }
+  }
 
+  private calculateHash(hashType: AssetHashType, assetHash?: string, bundling?: BundlingOptions): string {
     switch (hashType) {
       case AssetHashType.SOURCE:
-        return FileSystem.fingerprint(this.sourcePath, this.fingerprintOptions);
+        const sourceHash = FileSystem.fingerprint(this.sourcePath, this.fingerprintOptions);
+        return bundling
+          ? this.createBundlingHash(bundling, sourceHash)
+          : sourceHash;
       case AssetHashType.BUNDLE:
         if (!this.bundleDir) {
           throw new Error('Cannot use `AssetHashType.BUNDLE` when `bundling` is not specified.');
         }
         return FileSystem.fingerprint(this.bundleDir, this.fingerprintOptions);
       case AssetHashType.CUSTOM:
-        if (!props.assetHash) {
+        if (!assetHash) {
           throw new Error('`assetHash` must be specified when `assetHashType` is set to `AssetHashType.CUSTOM`.');
         }
         // Hash the hash to make sure we can use it in a file/directory name.
         // The resulting hash will also have the same length as for the other hash types.
-        return crypto.createHash('sha256').update(props.assetHash).digest('hex');
+        return bundling
+          ? this.createBundlingHash(bundling, assetHash)
+          : crypto.createHash('sha256').update(assetHash).digest('hex');
       default:
         throw new Error('Unknown asset hash type.');
     }
   }
+
+  private createBundlingHash(bundling: BundlingOptions, assetHash: string) {
+    const bundlingHashComponent = {
+      ...bundling,
+      image: bundling.image.imageHash,
```

Comment:
It feels a bit risky to serialize an object that includes non-data types (such as `BundlingDockerImage`) and then cherry-picking this specific object.

I wonder if it's safer to override `toJSON()` on `BundlingDockerImage` and just `JSON.stringify()` the whole thing?

Reply:
I've gone the `toJSON()` route and stringified the whole thing now.